cmd/test-server/zerotrace.go

```diff
@@ -0,0 +1,295 @@
+			break
+		}
+	}
+	return traceroute, nil
+}
+
+// setupPcap sets up the pcap handle on the required interface, and applies the filter and returns it
+func (z *zeroTrace) setupPcapAndFilter() (*pcap.Handle, error) {
+	pcapHdl, err := pcap.OpenLive(z.Iface, snaplen, promisc, time.Second)
+	if err != nil {
+		ErrLogger.Println("Handle error:", err)
+		return nil, err
+	}
+	if err = pcapHdl.SetBPFFilter(fmt.Sprintf("(tcp and port %d and host %s) or icmp", z.ClientPort, z.ClientIP)); err != nil {
+		ErrLogger.Fatal(err)
+		return nil, err
+	}
+	return pcapHdl, nil
+}
+
+// recvPackets listens on the provided pcap handler for packets sent, processes TCP and ICMP packets differently
+func (z *zeroTrace) recvPackets(pcapHdl *pcap.Handle, hops chan HopRTT) {
+	tempConn := z.Conn.(*tls.Conn)
+	tcpConn := tempConn.NetConn()
+
+	localSrcAddr := tcpConn.LocalAddr().String()
+	serverIP, _, _ := net.SplitHostPort(localSrcAddr)
+
+	z.IPIdHop = make(map[int][]SentPacketData)
+	packetStream := gopacket.NewPacketSource(pcapHdl, pcapHdl.LinkType())
+	var counter int
+
+	for packet := range packetStream.Packets() {
+		currTTL := z.CurrTTLIndicator
+		if packet == nil {
+			continue
+		}
+		ipLayer := packet.Layer(layers.LayerTypeIPv4)
+		tcpLayer := packet.Layer(layers.LayerTypeTCP)
+		icmpLayer := packet.Layer(layers.LayerTypeICMPv4)
+
+		if ipLayer != nil {
+			// To identify the TCP packet that we have sent
+			if tcpLayer != nil {
+				z.processTCPpkt(packet, serverIP)
+			}
+			// If it is an ICMP packet, check if it is the ICMP TTL exceeded one we are looking for
+			if icmpLayer != nil && counter != currTTL {
+				clientFound, err := z.processICMPpkt(packet, currTTL, &counter, hops)
+				if err == icmpPktError {
+					continue
+				} else if clientFound {
+					return
+				}
+			}
+		}
+		if counter > MaxTTLHops {
+			return
+		}
+	}
+}
+
+// sendTracePacket sets the ttlValue, sends the TTL limited probe on the tcpConn and return errors if any
+func (z *zeroTrace) sendTracePacket(ttlValue int) error {
+	tempConn := z.Conn.(*tls.Conn)
+	tcpConn := tempConn.NetConn()
+	ipConn := ipv4.NewConn(tcpConn)
+
+	rawBytes := []byte(stringToSend)
+	localSrcAddr := tcpConn.LocalAddr().String()
+	localSrcIP, localSrcPortString, _ := net.SplitHostPort(localSrcAddr)
+	localSrcPort, _ := strconv.Atoi(localSrcPortString)
+
+	ipLayer := &layers.IPv4{
+		Protocol: layers.IPProtocolTCP,
+		Version:  ipversion,
+		SrcIP:    net.ParseIP(localSrcIP),
+		DstIP:    net.ParseIP(z.ClientIP),
+	}
+	tcpLayer := &layers.TCP{
+		SrcPort: layers.TCPPort(localSrcPort),
+		DstPort: layers.TCPPort(z.ClientPort),
+		PSH:     true,
+		ACK:     true,
+	}
+	_ = tcpLayer.SetNetworkLayerForChecksum(ipLayer)
+
+	// Create the packet with the layers
+	buffer = gopacket.NewSerializeBuffer()
+
+	serializeErr := gopacket.SerializeLayers(buffer, options,
+		tcpLayer,
+		gopacket.Payload(rawBytes),
+	)
+	if serializeErr != nil {
+		ErrLogger.Println("Send Packet Error: Serialize: ", serializeErr)
```

Comment:
We could save ourselves all these log messages if the caller checks if `sendTracePacket` returned and error, and if so, we can do something like:
```Go
if err := z.sendTracePacket(ttlValue); err != nil {
  ErrLogger.Println("Send Packet Error: ", err)
  return traceroute, sendError
}
```

Reply:
Agreed, and done in ffe573f2b3312bb82a948fa5e36b54e97bd31cec